Fix: PreviewBridge mask editor KeyError with clipspace files + restore fresh image behavior

djdarcy opened the pull request:
On Windows 11 when using the PreviewBridge node's mask editor (accessed via right-click → "Open in Mask Editor"), two issues occurred in the latest ComfyUI build [v0.3.38](https://github.com/comfyanonymous/ComfyUI/releases/tag/v0.3.38) (commit [e6609dacdeeafa371fe4e9f303016a605a333a76](https://github.com/comfyanonymous/ComfyUI/commit/e6609dacdeeafa371fe4e9f303016a605a333a76)):

### Issue 1: Clipspace KeyError
1. Create and save a mask in the mask editor
2. First save works correctly 
3. Create and save a second mask → **KeyError** or **FileNotFoundError**
4. Subsequent mask edits would fail with errors like:
   ```
   KeyError: 'clipspace/clipspace-mask-1125217.3.png [input]'
   ```
### Issue 2: Lost "Fresh Start" Behavior  
5. After initial fix, new image generations would retain previous masks instead of starting with empty masks
6. But to maintain original behavior, expected new images to require creating new masks

## Reason for the new issues (from what I can gather) ...

### Clipspace Issue
The clipspace issue occurs due to a mismatch between ComfyUI's clipspace system and the Impact Pack's preview bridge system:

1. **Mask Editor Integration**: ComfyUI's mask editor creates files in the clipspace system with names like `clipspace-mask-12345.png [input]`
2. **Preview Bridge Expectation**: Impact Pack expects preview IDs like `$274-1748423590.3098016` 
3. **Frontend Conversion**: The JavaScript in `impact-image-util.js` is supposed to convert clipspace filenames to preview IDs via the `/impact/set/pb_id_image` API
4. **Timing Issue**: There's a race condition where the backend receives the clipspace filename before the frontend conversion completes

### Fresh Start Issue
The original code relied on cache invalidation to clear masks when images changed. My initial clipspace fix inadvertently made the cache more persistent, breaking the "fresh start" behavior where new images should begin with empty masks.

## Technical Details
The frontend JavaScript (`impact-image-util.js`) handles this conversion:
```javascript
Object.defineProperty(w, 'value', {
    async set(v) {
        if(v && v.constructor == String && v.startsWith('$')) {
            // from node feedback - uses preview ID system
        } else {
            // from clipspace - converts to preview ID  
            w._value = await loadImageFromUrl(image, node.id, v, false);
        }
    }
})
```

However, due to async timing, the backend sometimes receives the original clipspace filename instead of the converted preview ID.

## Solution

### Part 1: Clipspace Registration
Added a `register_clipspace_image()` method that:

1. **Detects clipspace files** by checking for "clipspace" or "[input]" in the filename
2. **Locates the actual file** by trying common clipspace paths
3. **Registers the file** in the preview bridge system using `core.set_previewbridge_image()`
4. **Creates dual registration** under both the preview ID and original clipspace path for compatibility

### Part 2: Fresh Start Behavior Restoration
Added explicit mask cache management:

1. **Track input changes** with `images_changed`/`latent_changed` flags
2. **Clear mask cache on change** by deleting `core.preview_bridge_last_mask_cache[unique_id]` when inputs change
3. **Preserve clipspace functionality** by only attempting clipspace registration when inputs haven't changed
4. **Respect restore_mask appropriately** - only for unchanged inputs, new inputs always start fresh

## Changes Made
- Added `register_clipspace_image()` static method to `PreviewBridge` class
- Added clipspace file detection logic in both `PreviewBridge.doit()` and `PreviewBridgeLatent.doit()`
- Enhanced cache validation to handle unregistered image parameters
- Added explicit mask cache clearing when images/latents change
- Modified restore_mask logic to respect "fresh start" behavior for new content
- Added safety check in `load_image()` to prevent KeyError when path lookup fails

## Behavior Matrix

| Scenario | Input Changed | Mask Cache | Result |
|----------|---------------|------------|---------|
| New generation | Yes | Cleared | Fresh mask (original behavior restored) |
| Same image, new clipspace mask | No | Preserved | Clipspace registration works |
| Same image, restore_mask="always" | No | Preserved | Mask restored as expected |
| Same image, restore_mask="never" | No | Preserved | Fresh mask |

## Did a few tests involving:
- First mask creation and save works
- Second mask creation and save works  
- Subsequent mask edits work without errors
- New image generations start with fresh/empty masks (original behavior)
- restore_mask settings work correctly for unchanged images
- Normal preview bridge functionality unchanged
- Backwards compatibility maintained

## Backwards Compatibility
This fix should theoretically be fully backwards compatible:
- No breaking changes to existing API
- Normal preview bridge workflow unchanged
- Only adds handling for the previously broken clipspace scenario
- Restores expected behavior for new image generations

## Things to consider...
For long-term stability, consider:

1. **Frontend Timing**: The async conversion in `impact-image-util.js` could be made more robust with proper promise chaining
2. **Unified System**: Eventually migrating to a single preview ID system throughout might eliminate this entire category of issues
3. **ComfyUI API Changes**: Not sure the best way to monitor ComfyUI updates that might affect clipspace/preview bridge interaction, but perhaps some sort of CI/CD?

This fix should theoretically address the immediate clipspace issue and the behavioral regression while maintaining full compatibility with existing workflows.

djdarcy committed:
fix: Handle clipspace files in PreviewBridge mask editor and restore fresh image behavior

- Fix KeyError when using mask editor multiple times
- Add register_clipspace_image() method to detect and register clipspace files
- Resolve timing issue between frontend JS conversion and backend processing
- Restore original "fresh start" behavior for new image generations
- Clear mask cache when images/latents change to ensure clean slate for new content
- Maintain backwards compatibility with existing preview bridge functionality
- Add dual registration for clipspace paths and preview IDs

Fixes issue where second+ mask saves would fail with:
KeyError: 'clipspace/clipspace-mask-XXXXX.png [input]'

Also fixed regression where new images would retain previous masks instead of starting fresh.

huanggou666 commented:
thanks ,solved my problem

djdarcy commented:
@huanggou666 Glad it worked! Let me know if you have any issues, so far it seems to be working as expected on my end.

ltdrdata commented:
Sorry for the delayed review of this PR.
Upon testing, there's an issue where `restore_mask` does not behave as intended.

Original behavior:

https://github.com/user-attachments/assets/58f3cab6-969a-4506-bed7-51c47c68bde9



This PR's behavior:

https://github.com/user-attachments/assets/136e1aca-fa67-48fb-8c14-65a3769b8703

djdarcy commented:
Now it's my turn to apologize, sorry didn't see your comment earlier. Here is a quick fix: [restore_mask_always_fix.patch](https://github.com/user-attachments/files/21357602/restore_mask_always_fix.patch). This should address the `restore_mask` issue, to apply the patch: `git apply restore_mask_always_fix.patch` from the Impact Pack project directory, or manually apply the 4 line changes shown in the diff. I'll do a couple tests later this evening, but it just adds a condition to prevent clearing the mask cache when `restore_mask="always"` and should adjust the restore logic to allow restoration with new images/latents for that setting. Cheers!

djdarcy committed:
Fix restore_mask 'always' and 'if_same_size' modes in PreviewBridge

- Fixed 'always' mode to properly preserve masks when changing input images/latents
- Fixed 'if_same_size' mode to preserve masks when switching between same-sized images
- Modified cache clearing logic to preserve cache for both 'always' and 'if_same_size' modes
- Updated restoration logic to handle size comparison correctly for 'if_same_size'

This addresses the issue where masks were being cleared even when restore_mask was set to 'always' or when switching between same-sized images with 'if_same_size'.

Fixes the issue discussed in PR #1009

djdarcy commented:
Alright finally had a moment to do the tests. Here is an update, latest changes at https://github.com/djdarcy/ComfyUI-Impact-Pack/commit/983f785ac995d383726334c1ede181eb1d98de82, I think this should cover all the cases you called out (`restore_mask="always"` and `restore_mask="if_same_size"` both appear to work now). 

Here is a quick vidcap for verification:

https://github.com/user-attachments/assets/8ab34576-1f2d-4bed-a394-f84a655984e6

Let me know if I missed anything!
